### ISSUE TITLE
[MNG-5493] Concurrency exception in Maven core while using parallel build

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6850]
**[Pengfei Zhang](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=superterry)** opened **[MNG-5493](https://issues.apache.org/jira/browse/MNG-5493?redirect=false)** and commented

My project uses parallel build for my "deployment" job in Hudson. From time to time, the build failed because of the following `ConcurrencyModificationException`. This apparently is not a plugin-in bug. Instead, it is a deeper issue in the core. I checked the source code, I think the "Properties" should be replaced by a concurrentHashMap. Accordingt to the java doc of "Properties", it cannot support concurrency.

```
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:2.4:single (make-assembly) on project ssm-viewer-spread: Execution make-assembly of goal org.apache.maven.plugins:maven-assembly-plugin:2.4:single failed.
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:225)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleThreadedBuilder$1.call(LifecycleThreadedBuilder.java:167)
	at org.apache.maven.lifecycle.internal.LifecycleThreadedBuilder$1.call(LifecycleThreadedBuilder.java:163)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:722)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution make-assembly of goal org.apache.maven.plugins:maven-assembly-plugin:2.4:single failed.
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:110)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	... 13 more
Caused by: java.util.ConcurrentModificationException
	at java.util.Hashtable$Enumerator.next(Hashtable.java:1200)
	at java.util.Hashtable.putAll(Hashtable.java:614)
	at org.apache.maven.project.DefaultProjectBuildingRequest.setSystemProperties(DefaultProjectBuildingRequest.java:166)
	at org.apache.maven.project.DefaultMavenProjectBuilder.toRequest(DefaultMavenProjectBuilder.java:79)
	at org.apache.maven.project.DefaultMavenProjectBuilder.buildFromRepository(DefaultMavenProjectBuilder.java:229)
	at org.apache.maven.project.DefaultMavenProjectBuilder.buildFromRepository(DefaultMavenProjectBuilder.java:251)
	at org.apache.maven.project.DefaultMavenProjectBuilder.buildFromRepository(DefaultMavenProjectBuilder.java:258)
	at org.apache.maven.plugin.assembly.archive.task.AddDependencySetsTask.addDependencySet(AddDependencySetsTask.java:162)
	at org.apache.maven.plugin.assembly.archive.task.AddDependencySetsTask.execute(AddDependencySetsTask.java:124)
	at org.apache.maven.plugin.assembly.archive.phase.DependencySetAssemblyPhase.execute(DependencySetAssemblyPhase.java:76)
	at org.apache.maven.plugin.assembly.archive.DefaultAssemblyArchiver.createArchive(DefaultAssemblyArchiver.java:183)
	at org.apache.maven.plugin.assembly.mojos.AbstractAssemblyMojo.execute(AbstractAssemblyMojo.java:436)
	at org.apache.maven
```

---

**Affects:** 3.0.3, 3.0.5

**Issue Links:**
- [MNG-5670](https://issues.apache.org/jira/browse/MNG-5670) ConcurrentModificationException during DefaultMaven.newRepositorySession
 (_**"duplicates"**_)

1 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

Please read https://cwiki.apache.org/confluence/display/MAVEN/Submitting+Issue+for+Maven's+Core

The POMs are all read before the build executes concurrently. So it's hard to diagnose without a sample. If the assembly plugin is modifying these properties at runtime that's probabably not good.

**Comment by jira-importer** [comment]
**[Mike Beaubien](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mbeaubie)** commented

We're facing a similar issue, we see it when doing parallel builds, here's the relevant stack trace:

Caused by: java.util.ConcurrentModificationException
at java.util.Hashtable$Enumerator.next(Hashtable.java:1200)
at java.util.Hashtable.putAll(Hashtable.java:614)
at org.apache.maven.project.artifact.MavenMetadataSource.getSystemProperties(MavenMetadataSource.java:757)
at org.apache.maven.project.artifact.MavenMetadataSource.retrieveRelocatedProject(MavenMetadataSource.java:575)
at org.apache.maven.project.artifact.MavenMetadataSource.retrieve(MavenMetadataSource.java:191)
at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse(DefaultLegacyArtifactCollector.java:532)
at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse(DefaultLegacyArtifactCollector.java:584)
at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.collect(DefaultLegacyArtifactCollector.java:144)
at org.apache.maven.artifact.resolver.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:494)
at org.apache.maven.artifact.resolver.DefaultArtifactResolver.resolveWithExceptions(DefaultArtifactResolver.java:350)
at org.apache.maven.artifact.resolver.DefaultArtifactResolver.resolveTransitively(DefaultArtifactResolver.java:344)
at org.apache.maven.artifact.resolver.DefaultArtifactResolver.resolveTransitively(DefaultArtifactResolver.java:319)
at org.apache.maven.artifact.resolver.DefaultArtifactResolver.resolveTransitively(DefaultArtifactResolver.java:288)
at org.apache.maven.plugin.assembly.artifact.DefaultDependencyResolver.resolveTransitively(DefaultDependencyResolver.java:200)
at org.apache.maven.plugin.assembly.artifact.DefaultDependencyResolver.resolve(DefaultDependencyResolver.java:122)
at org.apache.maven.plugin.assembly.archive.DefaultAssemblyArchiver.createArchive(DefaultAssemblyArchiver.java:177)
at org.apache.maven.plugin.assembly.mojos.AbstractAssemblyMojo.execute(AbstractAssemblyMojo.java:436)
at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:133)
... 14 more

I looked at the source, this is caused because org.apache.maven.project.artifact.MavenMetadataSource.getSystemProperties is copying System.getProperties() into a new Properties object. The result of System.getProperties() is not cloned, so if another thread gets in there and adds something to the System properites during the copy, you get the exception.

I'd recommend you go and catch the concurrent mod exception here and re-try. I'll submit a patch in the meantime.

Thanks,
Mike

**Comment by jira-importer** [comment]
**[jwalton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jwalton)** commented

Looks like this was duplicated, and fixed, as MNG-5670 ([8980f67b9b1d](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commitdiff;h=8980f67b9b1d0dbc06b68f74a88b9f8589c40e23;hp=e3000a09c94396d7fd7d65fc236904aa07c7f7f2) in 3.2.3).
